Add hotkey for narrowing to what is in the compose box

mahimg opened the pull request:
Currently, the hotkey is set to meta+I. Shortcut needs to be finalized. 
Node tests to be added.
Fixes: #6546

mahimg commented:
@showell @rishig If the code looks fine then I can proceed with the node tests.

lonerz commented on static/js/hotkey.js:
```diff
@@ -40,6 +40,10 @@ var keydown_unshift_mappings = {
 };
 
+var keydown_meta_mappings = {
+    73: {name: 'narrow_compose', message_view_only: false}, // 'I'
```
comment should be lowercase `i`

mahimg commented:
All done here. Finalizing the trigger key is left. Requesting review @showell @lonerz

showell commented on static/js/hotkey.js:
```diff
@@ -532,6 +544,23 @@ exports.process_hotkey = function (e, hotkey) {
         }
 
+        if (event_name === "narrow_compose" && compose_state.composing()) {
+            if (compose_state.get_message_type() === 'stream' && compose_state.stream_name() !== '') {
```
Can we move most of this code into a function called `narrow.to_compose_target`?  We try to minimize the amount of code in `hotkey.js` so that it only dispatches stuff.

showell commented on static/js/narrow.js:
```diff
@@ -265,6 +265,49 @@ exports.activate = function (raw_operators, opts) {
+        select_first_unread: true,
+    };
+
+    if (invalid_recipients.length !== 0 || recipient === '') {
```
It's usually good style to do early-return for the "error" condition, not the "normal" condition.  And then it's not clear to me what the rest of this function is supposed to do.

mahimg replied:
I will add some comments here.

showell commented on frontend_tests/node_tests/hotkey.js:
```diff
@@ -402,4 +410,52 @@ function stubbing(func_name_to_stub, test_function) {
     overlays.is_active = return_false;
     overlays.drafts_open = return_false;
+
+    assert_unmapped('narrow_compose', false, false, true);
```
We can greatly simplify the hotkey tests by just having a single function in narrow.js that does all the heavy lifting.  If you read the rest of the hotkey code, you'll see that most keys just dispatch to a single function in some other module that does all the heavy lifting.

showell commented:
@mahimg Thanks for working on this!  I made several specific comments, but I think the high level concerns here are about code organization:

* Keep hotkey.js simple, and let `narrow.js` do all the heavy lifting.
* See if there are existing functions in `narrow.js` that we can leverage better.
* Avoid unneeded code (or if code is needed for edge cases, add clear comments).

mahimg commented on static/js/hotkey.js:
```diff
@@ -532,6 +544,23 @@ exports.process_hotkey = function (e, hotkey) {
+                var stream = compose_state.stream_name();
+                var topic = compose_state.subject();
+                narrow.by_stream_topic_name(stream, topic);
+                compose_actions.start("stream");
```
After narrowing to the required page the compose box closes. Compose box remains open only for pm-with narrow.

showell replied:
We should figure out a way not to close the compose box in the `narrow` code, then, because closing and then re-opening may cause flickering.  There may be ways to do this with the existing functions, or, if not, we should refactor accordingly in a prep commit.

mahimg commented:
Thanks for the review @showell 
I have shifted most of the code to `narrow.js`. Also, added some comments on the edge cases.
`narrow.js` now has functions which can directly narrow to a given stream/topic and pm, with proper handling of edge cases too.
Node tests in hotkey are simple now, most of the changes are because of the inclusion of 'meta+' hotkey.
I have done rigorous testing on the functions in `narrow.js` to keep up with the edge cases.

I have refactored the function `compose_actions.on_narrow()` to keep the compose box open if it was open before the narrow call.

showell commented:
@gnprice @rishig The code here LGTM.  We'll want to add this to keyboard help.  I think this is a sufficiently risky change that we'll want to test-deploy it on czo first.

showell commented:
@mahimg The test failure here is apparently due to a Casper flake.  Can you add meta-I to the documentation and squash this all into one commit?

showell commented:
@mahimg I also had trouble testing with FF ubuntu, because it's not clear what key this maps to.

mahimg commented:
@showell Yes, firefox has stopped recognizing windows key as metaKey from version 48. I left a few comments on chat too. We may want to shift this to AltKey maybe.

timabbott commented:
Closing in favor of the reworked version at #9511.
